Cython/Utility/MemoryView_C.c

```diff
@@ -26,8 +26,58 @@ typedef struct {
 #endif
 
 #define __pyx_atomic_int_type int
+#define __pyx_nonatomic_int_type int
+
+// for standard C/C++ atomics, get the headers first so we have ATOMIC_INT_LOCK_FREE
+// defined when we decide to use them
+#if CYTHON_ATOMICS && (defined(__STDC_VERSION__) && \
+                        (__STDC_VERSION__ >= __STDC_VERSION__) && \
+                        !defined(__STDC_NO_ATOMICS__))
+    #include <stdatomic.h>
+#elif CYTHON_ATOMICS && (defined(__cplusplus) && ( \
+                    (__cplusplus >= 201103L) || \
+                    (defined(_MSC_VER) && _MSC_VER >= 1700)))
```

Comment:
I've added a comment.

I'm inclined not to get into using `<stdatomic.h>` in C++ mode. Everything I've seen suggests that it's very compiler-specific and inconsistent. And I'm trying to get away from compiler-specific and inconsistent as much as possible here

Reply:
> For reference, even VS 2017 (another "17", yay!) is already EOL, so as soon as Python 2.7 is dropped (and thus VS2008), this could be made unconditional. Surely Cython will want to have conservative lower bounds for compiler versions, but older than VS2017 (or maybe even VS2019) makes no sense IMO.

Possibly. There is a non-atomic fallback of course (and it looks like it's been used for years on Windows) so in this case Cython could support any users choosing to use really odd/old compilers with the fallback.

> In case it's of interest, this is the [page](https://scipy.github.io/devdocs/dev/toolchain.html) where scipy documents the relevant constraints for itself (that I help maintain) - as of 1.9, it requires VS2019 (with the vc142 toolchain).

That was useful. The slight difference is that the majority of Scipy users probably download pre-compiled versions while compiling your own code is the core feature of Cython. So it needs to be a little more tolerant.